File: src/main/java/zaftnotameni/creatania/ponder/CobblegenScenes.java
package zaftnotameni.creatania.ponder;

import com.simibubi.create.foundation.ponder.SceneBuilder;
import com.simibubi.create.foundation.ponder.SceneBuildingUtil;
import net.minecraft.core.Direction;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraftforge.registries.ForgeRegistries;
import org.jetbrains.annotations.NotNull;
import zaftnotameni.creatania.recipes.cobblegen.CobblegenRecipe;

import static net.minecraft.world.level.block.LiquidBlock.LEVEL;

public class CobblegenScenes {

  public CobblegenRecipe recipe;

  public CobblegenScenes(CobblegenRecipe r) { recipe = r; }

  public void makeCobblegenScene(SceneBuilder s, SceneBuildingUtil u) {
    var x = new CreataniaPonderUtils(s, u);
    var outputItem = recipe
      .getResultItem()
      .getItem();
    x.scene.title("cobblegen_" + outputItem
      .getRegistryName()
      .getPath(), "Magic Cobblegen");
    x.scene.showBasePlate();
    x.scene.scaleSceneView(1.5f);

    for (var i = 0; i < 2; i++) for (var j = 0; j < 2; j++) {
      var even = (i + j) % 2 == 0;
      var block = even ? Blocks.BLACK_CONCRETE : Blocks.GRAY_CONCRETE;
      var y = x.at(2 - i, 1, 2 - j);
      x.scene.world.setBlock(y.asBlockPos(), block.defaultBlockState(), false);
      x.scene.world.showSection(y.asSelection(), Direction.DOWN);
      x.scene.idle(10);
    }

    var fluidA = x.at(0, 1, 2);
    var flowA = x.at(0, 1, 1);

    var fluidB = x.at(2, 1, 0);
    var flowB = x.at(1, 1, 0);
    var result = x.at(0, 1, 0);

    var outputBlock = ForgeRegistries.BLOCKS.getValue(outputItem.getRegistryName());

    showFlowingFluid(x,0, fluidA, flowA);
    showFlowingFluid(x,1, fluidB, flowB);

    x.scene.world.setBlock(result.asBlockPos(), outputBlock.defaultBlockState(), true);
    x.scene.world.showSection(result.asSelection(), Direction.UP);
    x.keyFrameText("Blocks generated with mana!", result.top(), 60).run();
  }

  private void showFloorPiece(CreataniaPonderUtils x, int i) {
    var block = i % 2 == 0 ? Blocks.SNOW.defaultBlockState() : Blocks.WHITE_CONCRETE.defaultBlockState();
    x.scene.world.setBlock(x.at(i, 0, 0).asBlockPos(), block, false);
    x.scene.idle(5);
    x.scene.world.showSection(x.at(i, 0, 0).asSelection(), Direction.UP);
    x.scene.idle(5);
  }

  private void showFlowingFluid(CreataniaPonderUtils x, int pIndex, CreataniaPonderUtils.XYZ fluid, CreataniaPonderUtils.XYZ flow) {
    x.scene.world.setBlock(fluid.asBlockPos(), blockOf(pIndex), false);
    x.scene.world.setBlock(flow.asBlockPos(), blockOf(pIndex), false);

    x.scene.world.showSection(fluid.asSelection(), Direction.DOWN);
    x.scene.idle(40);
    x.scene.world.showSection(flow.asSelection(), Direction.DOWN);

    var requiresFullBlock = recipe.getInputFluidStacks().get(pIndex).getRequiredAmount() > 999;
    if (!requiresFullBlock) {
      x.scene.world.modifyBlock(flow.asBlockPos(), bs -> bs.setValue(LEVEL, 5), false);
      x.keyFrameText("Flowing fluids will NOT be consumed in the process.", flow.top(), 50).run();
    } else {
      x.keyFrameText("Source fluids WILL be consumed in the process.", flow.top(), 50).run();
    }

    x.scene.idle(60);
    if (requiresFullBlock) {
      x.scene.world.hideSection(flow.asSelection(), Direction.DOWN);
    }
  }

  @NotNull private BlockState blockOf(int pIndex) {
    return recipe
      .getInputFluidStacks()
      .get(pIndex)
      .getMatchingFluidStacks()
      .get(0)
      .getFluid()
      .defaultFluidState()
      .createLegacyBlock();
  }

}

File: src/main/java/zaftnotameni/creatania/ponder/CobblegenScenes.kt
package zaftnotameni.creatania.ponder

import com.simibubi.create.foundation.ponder.SceneBuilder
import com.simibubi.create.foundation.ponder.SceneBuildingUtil
import net.minecraft.core.Direction
import net.minecraftforge.registries.ForgeRegistries
import zaftnotameni.creatania.recipes.cobblegen.CobblegenRecipe

class CobblegenScenes(var recipe : CobblegenRecipe) {
  fun header(s : SceneBuilder?, u : SceneBuildingUtil?) : CreataniaPonderUtils {
    val x = CreataniaPonderUtils(s, u)
    val outputItem = recipe.resultItem.item
    x.scene.title("cobblegen_" + outputItem.registryName!!.path, "Magic Cobblegen")
    x.scene.showBasePlate()
    return x
  }

  fun makeCobblegenFlowingSourceScene(s : SceneBuilder?, u : SceneBuildingUtil?) {
    val c = header(s, u)
    checkerboard(c, c.at(0, 1, 1), c.at(2, 1, 1))
    val fs = recipe.asFlowingSource()
    val flowing = fs.a
    val source = fs.b
    fluidWithLevelAt(c, 9, flowing, 2, 1, 0)
    val ff = fluidWithLevelAt(c, 5, flowing, 1, 1, 0)
    c.keyFrameText("Flowing " + nameOf(flowing) + " will not be consumed", ff!!.center(), 60).run()
    c.scene.idle(60)
    val ss = fluidWithLevelAt(c, 9, source, 0, 1, 0)
    c.keyFrameText(nameOf(source) + " Source will be consumed", ss!!.center(), 60).run()
    c.scene.idle(60)
    val output = c.at(0, 1, 0)
    val block = ForgeRegistries.BLOCKS.getValue(recipe.resultItem.item.registryName)
    c.scene.world.setBlock(output.asBlockPos(), block!!.defaultBlockState(), true)
    c.keyFrameText(block.name.string + " will be generated", output.center(), 60).run()
  }

  fun makeCobblegenFlowingFlowingScene(s : SceneBuilder?, u : SceneBuildingUtil?) {
    val c = header(s, u)
    checkerboard(c, c.at(1, 1, 1), c.at(1, 1, 1))
    val ff = recipe.asFlowingFlowing()
    val a = ff.a
    val b = ff.b
    fluidWithLevelAt(c, 9, a, 2, 1, 0)
    val aa = fluidWithLevelAt(c, 5, a, 1, 1, 0)
    c.keyFrameText("Flowing " + nameOf(a) + " will not be consumed", aa!!.center(), 60).run()
    c.scene.idle(60)
    fluidWithLevelAt(c, 9, b, 0, 1, 2)
    val bb = fluidWithLevelAt(c, 5, b, 0, 1, 1)
    c.keyFrameText("Flowing " + nameOf(b) + " will not be consumed", bb!!.center(), 60).run()
    c.scene.idle(60)
    val output = c.at(0, 1, 0)
    val block = ForgeRegistries.BLOCKS.getValue(recipe.resultItem.item.registryName)
    c.scene.world.setBlock(output.asBlockPos(), block!!.defaultBlockState(), true)
    c.scene.world.showSection(output.asSelection(), Direction.UP)
    c.keyFrameText(nameOf(block) + " will be generated", output.center(), 60).run()
  }

  fun makeCobblegenScene(s : SceneBuilder?, u : SceneBuildingUtil?) {
    if (recipe.hasSource()) {
      makeCobblegenFlowingSourceScene(s, u)
    } else makeCobblegenFlowingFlowingScene(s, u)
  }
}